bigtable/bigtable.go

```diff
@@ -52,17 +55,84 @@ const mtlsProdAddr = "bigtable.mtls.googleapis.com:443"
 //
 // A Client is safe to use concurrently, except for its Close method.
 type Client struct {
-	connPool          gtransport.ConnPool
-	client            btpb.BigtableClient
-	project, instance string
-	appProfile        string
+	connPool             gtransport.ConnPool
+	client               btpb.BigtableClient
+	project, instance    string
+	appProfile           string
+	metricsTracerFactory *builtinMetricsTracerFactory
 }
 
 // ClientConfig has configurations for the client.
 type ClientConfig struct {
 	// The id of the app profile to associate with all data operations sent from this client.
 	// If unspecified, the default app profile for the instance will be used.
 	AppProfile string
+
+	// If not set, client side metrics will be collected and exported
+	//
+	// To disable client side metrics, set 'MetricsProvider' to 'NoopMetricsProvider'
+	//
+	// To use a custom meter provider,
+	// 1) use 'WithBuiltIn' to get built-in metrics meter provider options
+	// 2) create meter provider
+	// 3) Set 'MetricsProvider' to the created meter provider
+	MetricsProvider MetricsProvider
+}
+
+// MetricsProvider is a wrapper for built in metrics meter provider
+type MetricsProvider interface {
+	isMetricsProvider()
+}
+
+// NoopMetricsProvider can be used to disable built in metrics
+type NoopMetricsProvider struct{}
+
+func (NoopMetricsProvider) isMetricsProvider() {}
+
+// CustomOpenTelemetryMetricsProvider can be used to collect and export builtin metric with custom meter provider
+type CustomOpenTelemetryMetricsProvider struct {
+	MeterProvider *sdkmetric.MeterProvider
+}
+
+func (CustomOpenTelemetryMetricsProvider) isMetricsProvider() {}
+
+/*
+BuiltInMeterProviderOptions returns built in metrics periodic reader with Google Cloud Monitoring exporter
+
+This functionality is needed for cases where end users want to export metrics to their own metrics
+aggregator in addition to the Google Cloud Monitoring. To accompish this, end users would use this
+method to create options to be passed in to the OpenTelemretry which in turn gets passed in as an
+option to bigtable client for example:
+
+	builtInMeterProviderOpts, err := BuiltInMeterProviderOptions(project)
+	if err != nil {
+		// Handle error
+	}
+	meterProviderOpts := append(builtInMeterProviderOpts, usersMeterProviderOpts...)
+
+	// Create custom meter provider which exports to user's metrics aggregator and GCM
+	customMeterProvider = sdkmetric.NewMeterProvider(meterProviderOpts...)
+
+	clientConfig := ClientConfig{
+		MetricsProvider: CustomOpenTelemetryMetricsProvider{
+			MeterProvider: customMeterProvider,
+		},
+	}
+
+	client, err := NewClientWithConfig(context.Background(), project, instance, clientConfig)
+*/
+func BuiltInMeterProviderOptions(project string) ([]sdkmetric.Option, error) {
```

Comment:
I would be wary of putting a third-party project into the bigtable public API. This means if that package takes a breaking change in a minor version for whatever reason it is now a breaking change for bigtable. Also if otel ever released a v2 of their module this function is tightly coupled to the v1 implementation. Could this code live somewhere else?

Reply:
I could put it as a repository in googleapis or copy the whole thing into a third party folder in the bigtable client itself

Reply:
Removed the public API. Will add it later with modifications

Reply:
Added it to: https://github.com/googleapis/google-cloud-go/pull/10629